# CAPSTONE PROJECT: GENRE CLASSIFICATION 

### Genre Consolidation 
Stephanie Barrett | September 5, 2023

This project aims to use machine learning to classify songs into the correct genre based on particular attributes. 

**Introduction:** This is notebook 2 of 4 that was used in the genre classification project.
This notebook contains further exploration of the genres in our spotify dataset. In this notebook, we remove a total of 28 genres and reduce the total number of unique genres to 85. The new DataFrame with 85 genres is saved to another csv file that will be used in the next notebook: modeling. 

***
**TABLE OF CONTENTS**

[Introduction](#Genre-Consolidation)

[Sleep](#Sleep)

[Study](#Study)

[Ambient](#Ambient)

[Sad & Happy](#Sad-&-Happy)

[Languages](#Languages)

[Next Steps](#NEXT-STEPS)

***

In [1]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt 
import seaborn as sns 

In [2]:
# Reading in our cleaned dataset 
spotify_cleaned = pd.read_csv('~/Desktop/CapstoneProject/data/spotify_cleaned_final.csv', index_col=0)

In [3]:
# Making a copy - we will eventually save this DataFrame to a csv file that we can use for modeling. 
consolidated_spotify = spotify_cleaned.copy()

In [4]:
# Transforming the 'explicit' column from bool to binary
consolidated_spotify['explicit'] = consolidated_spotify['explicit'].astype(int)

In [5]:
# Sanity check 
consolidated_spotify['explicit'].unique()

array([0, 1])

In [6]:
# Looking at our unique genres 
consolidated_spotify['track_genre'].unique()

array(['acoustic', 'afrobeat', 'alt-rock', 'alternative', 'ambient',
       'anime', 'black-metal', 'bluegrass', 'blues', 'brazil',
       'breakbeat', 'british', 'cantopop', 'chicago-house', 'children',
       'chill', 'classical', 'club', 'comedy', 'country', 'dance',
       'dancehall', 'death-metal', 'deep-house', 'detroit-techno',
       'disco', 'disney', 'drum-and-bass', 'dub', 'dubstep', 'edm',
       'electro', 'electronic', 'emo', 'folk', 'forro', 'french', 'funk',
       'garage', 'german', 'gospel', 'goth', 'grindcore', 'groove',
       'grunge', 'guitar', 'happy', 'hard-rock', 'hardcore', 'hardstyle',
       'heavy-metal', 'hip-hop', 'honky-tonk', 'house', 'idm', 'indian',
       'indie-pop', 'indie', 'industrial', 'iranian', 'j-dance', 'j-idol',
       'j-pop', 'j-rock', 'jazz', 'k-pop', 'kids', 'latin', 'latino',
       'malay', 'mandopop', 'metal', 'metalcore', 'minimal-techno', 'mpb',
       'new-age', 'opera', 'pagode', 'party', 'piano', 'pop-film', 'pop',
       'pow

In [7]:
# Total number of genres 
consolidated_spotify['track_genre'].nunique()

113

We are going to start off by removing some genres of music that are probably a combination of noise (like nature sounds, white noise, etc.) and "soothing" or "relaxing" songs that overlap with other genres present in our dataset. 
- ambient 
- study 
- sleep 
- sad 
- happy

Let's check them out just to be sure. 

## Sleep 

In [8]:
# All songs from the 'sleep' genre
sleep = consolidated_spotify.loc[consolidated_spotify['track_genre'] == 'sleep']
sleep

track_id  \
101000  07UDTaRYJAsIhUZTyZSUzM   
101001  0CMYUXTTTmI6Lwc0opH2XG   
101002  5LEHlRPjGcZ5RdagAwXpHS   
101003  58xXb7ASilQ9WYtninubrY   
101004  6kOJ0Ylenwr6Up0g4lDORg   
...                        ...   
101994  5BVK02c2QI9ODFm8liVSdw   
101996  3OwZ8e7lb5Assv14ZRW1rU   
101997  3QZrgBLusdGcYJtST0ucXX   
101998  6hD4qRIXuuD2cM7vX6eV23   
101999  4CyVr80Xs1y26QtN7IImuc   

                                                 artists  \
101000                                              Huma   
101001                                        Sohn Aelia   
101002                                   Little Circuits   
101003                            White Noise for Babies   
101004                            White Noise for Babies   
...                                                  ...   
101994  Rain Sounds;Spa & Spa;Nature Sounds Nature Music   
101996                                       Rain Sounds   
101997                                       Rain Sounds   
101998                                    Meeresrauschen   
101999                                    Meeresrauschen   

                                               album_name  \
101000                                                Ons   
101001                                      Herinneringen   
101002                              Beautiful Imagination   
101003                                     Thinned Phaser   
101004                                         Rising Sun   
...                                                   ...   
101994                  18 Relaxing Rain Sounds - No Fade   
101996                                    Soft Rain Sleep   
101997                                   White Noise Rain   
101998  Meeresrauschen zur Entspannung, für Meditation...   
101999  Meeresrauschen zur Entspannung, für Meditation...   

                           track_name  popularity  duration_ms  explicit  \
101000                            Ons          71       193239         0   
101001                  Herinneringen          74       172357         0   
101002          Beautiful Imagination          71       143052         0   
101003                           Aqua           0       157024         0   
101004                     Rising Sun           0       180000         0   
...                               ...         ...          ...       ...   
101994  10 Minute Loopable Heavy Rain          32       591807         0   
101996                   Forrest Rain          32       102649         0   
101997               White Noise Rain          32       173608         0   
101998                          Relax          32       120353         0   
101999                       Chillout          32       135421         0   

        danceability    energy  key  loudness  mode  speechiness  \
101000        0.1110  0.018500    0   -32.335     1       0.0424   
101001        0.0685  0.003420    0   -37.156     1       0.0418   
101002        0.1020  0.033300    3   -34.177     1       0.0476   
101003        0.2400  0.011900    6   -33.515     0       0.0760   
101004        0.1350  0.000804    4   -34.772     1       0.0507   
...              ...       ...  ...       ...   ...          ...   
101994        0.1080  0.774000   10   -17.200     0       0.0725   
101996        0.4680  0.993000   10   -24.274     0       0.0386   
101997        0.1740  0.895000    2   -24.459     1       0.0468   
101998        0.1880  0.999000   10   -18.900     1       0.0510   
101999        0.1970  0.998000    9   -18.373     1       0.0627   

        acousticness  instrumentalness  liveness  valence    tempo  \
101000         0.957             0.940     0.087  0.04590   76.153   
101001         0.990             0.910     0.106  0.05810   73.644   
101002         0.797             0.927     0.111  0.05910   89.605   
101003         0.995             0.828     0.103  0.03580  138.009   
101004         0.968             0.728     0.108  0.03460  141.930   
...      

There is a total of 945 tracks. Already from glancing at the top and bottom, we can see track and album titles like "White Noise for Babies", "Rain Sounds." Let's check out a random sample of rows. 

In [9]:
# 100 random rows from sleep genre 
sleep.sample(100)

track_id  \
101934  5dYGrOlxDiibQ2qiVe8aNI   
101180  42nq6tG2mUQQhkBIM0kPuu   
101626  6cONGJWwrp7KmgmLMjSDVX   
101166  3InJfiPB0ZOw9xg0Z9BV7G   
101905  0IsEyOQ5ObJcX1tnuXeXqX   
...                        ...   
101171  2H03nIdKechb09iRBKaZS6   
101087  1umCMBSrqMBqmwNl1VfYus   
101009  6VwRN2gV7xF8pPxwry5Nvz   
101379  043f7kIHqTGBaOtN2tmZqB   
101876  5b87m5IILakknzxi0kuWHg   

                                                  artists  \
101934                         Baby Sleep Lullaby Academy   
101180  Traditional;London Philharmonic Orchestra;The ...   
101626                                         Pink Noise   
101166           Traditional;Royal Philharmonic Orchestra   
101905                                         Rnwy Lites   
...                                                   ...   
101171           Traditional;Royal Philharmonic Orchestra   
101087                                      Nature Sounds   
101009                             White Noise for Babies   
101379                         Nature Sounds Nature Music   
101876                             Brown Noise Baby Sleep   

                                               album_name  \
101934  Sleep Music to Help You Relax All Night: New A...   
101180                        Classical Christmas, Vol. 2   
101626                             Pink Noise Rain Sounds   
101166                        Classical Christmas, Vol. 1   
101905                             Breathe into the Still   
...                                                   ...   
101171                        Classical Christmas, Vol. 1   
101087                               Relaxing Rain Sounds   
101009                                     Twilight Noise   
101379  Tropical Storm for Deep Sleep - Thunderstorm S...   
101876                             Brown Noise For Babies   

                                              track_name  popularity  \
101934                         Meditation for the Dreams          32   
101180         The Trumpet Shall Sound (Remastered 2014)           0   
101626                         Calm Rain Water (No Fade)          35   
101166                    O Holy Night (Remastered 2014)           0   
101905                            Breathe into the Still          59   
...                                                  ...         ...   
101171                Away in a Manger (Remastered 2014)           0   
101087                                     Serenity Rain           0   
101009                                        Soft Light           0   
101379  Tropical Thunder healing Sounds of Mother Nature          39   
101876                            Baby Sleep Brown Noise          32   

        duration_ms  explicit  danceability  energy  key  loudness  mode  \
101934       209142         0        0.1450  0.0395    7   -28.457     1   
101180       262663         0        0.3730  0.2890    2   -11.743     1   
101626        66000         0        0.1290  0.5670    1   -34.024     1   
101166       163698         0        0.1630  0.3420    4   -11.015     1   
101905       192000         0        0.0667  0.0895    0   -30.760     1   
...             ...       ...           ...     ...  ...       ...   ...   
101171       216314         0        0.1120  0.2150    9   -17.766     1   
101087       246334         0        0.4970  0.9880    1   -26.516     1   
101009       186707         0        0.3340  0.2650    2   -13.579     1   
101379       360000         0        0.1640  0.5470   10   -20.848     0   
101876       120000         0        0.1350  0.1200    1   -38.378     1   

        speechiness  acousticness  instrumentalness  liveness  valence  \
101934       0.0403      0.924000           0.98200     0.103  0.03440   
101180       0.0384      0.935000           0.59400     0.547  0.26200   
101626       0.0523      0.081200           0.92300     0.363  0.03380   
101166       0.0371      0.980000           0.00124     0.110  0.08150   

This confirms my thought above. In addition to white noise and nature sounds we also see a mix of classical music. 

In [10]:
# Songs of classical genre
classical = consolidated_spotify.loc[consolidated_spotify['track_genre'] == 'classical']

In [11]:
# Shape of this genre 
classical.shape

(759, 20)

We have a descent amount of tracks in our classical genre. I see no need to move any of these songs to more "appropriate" genres. Let's go ahead and remove this genre, `sleep` from our data set. 

In [12]:
sleep_index = list(sleep.index)

In [13]:
# Removing any song with 'sleep' as a genre from the DataFrame
consolidated_spotify.drop(sleep_index, inplace=True)

In [14]:
# Sanity check 
consolidated_spotify['track_genre'].unique()

array(['acoustic', 'afrobeat', 'alt-rock', 'alternative', 'ambient',
       'anime', 'black-metal', 'bluegrass', 'blues', 'brazil',
       'breakbeat', 'british', 'cantopop', 'chicago-house', 'children',
       'chill', 'classical', 'club', 'comedy', 'country', 'dance',
       'dancehall', 'death-metal', 'deep-house', 'detroit-techno',
       'disco', 'disney', 'drum-and-bass', 'dub', 'dubstep', 'edm',
       'electro', 'electronic', 'emo', 'folk', 'forro', 'french', 'funk',
       'garage', 'german', 'gospel', 'goth', 'grindcore', 'groove',
       'grunge', 'guitar', 'happy', 'hard-rock', 'hardcore', 'hardstyle',
       'heavy-metal', 'hip-hop', 'honky-tonk', 'house', 'idm', 'indian',
       'indie-pop', 'indie', 'industrial', 'iranian', 'j-dance', 'j-idol',
       'j-pop', 'j-rock', 'jazz', 'k-pop', 'kids', 'latin', 'latino',
       'malay', 'mandopop', 'metal', 'metalcore', 'minimal-techno', 'mpb',
       'new-age', 'opera', 'pagode', 'party', 'piano', 'pop-film', 'pop',
       'pow

We no longer have 'sleep' as a genre. Let's move onto study music. 

 ## Study 


In [15]:
# All songs from the genre 'study'
study = consolidated_spotify.loc[consolidated_spotify['track_genre'] == 'study']

In [16]:
study

track_id                      artists  \
105000  6qbe2xJVwt7wwpeocsrVvc            Talented Mr Tipsy   
105001  5ZOwkvgMAeI0G2USq0cIFQ                        O k O   
105002  2L2aSrua8ZBRMogFhF6I1Q                        4to28   
105003  1tNwm7phOIMzGC1VO9Dggr                  Joi Casette   
105004  4wkQzdybc3qK4nNaxHLzvO                    hope mona   
...                        ...                          ...   
105995  0p9HloqJbkgokaktOhiwkn  Sarah, the Illstrumentalist   
105996  2RpHErrNFoo8A7MKAQS7tD                        DGHTR   
105997  2nHUre6dQezPvxBkeTAba5  Sarah, the Illstrumentalist   
105998  63QCKDyQB7OoiWOfwWkysb  Sarah, the Illstrumentalist   
105999  30W2HFDJ03F2U5j1JdzFfq                   Matt Large   

                            album_name         track_name  popularity  \
105000                          Ritual             Ritual          53   
105001                       neomatter          neomatter          49   
105002                         Elastic            Elastic          50   
105003                    One Two Tree       One Two Tree          47   
105004                          humify             humify          47   
...                                ...                ...         ...   
105995                         FLOWERS          Chamomile           6   
105996               HoriZon Butterfly          Red Maple           2   
105997                  Constellations        Capricornus           6   
105998  Pocket Full of Crystals: Vol 2      Orange Jasper           6   
105999              As Real As It Gets  Mean Street Blues           6   

        duration_ms  explicit  danceability  energy  key  loudness  mode  \
105000       173333         0         0.657   0.717   10    -6.007     0   
105001       104433         0         0.670   0.371    4   -13.830     0   
105002       121600         0         0.736   0.410    5   -14.587     0   
105003       121875         0         0.797   0.263    6   -13.089     0   
105004       161032         0         0.496   0.198    3   -18.740     1   
...             ...       ...           ...     ...  ...       ...   ...   
105995       140199         0         0.640   0.296    3    -9.142     0   
105996       119111         0         0.814   0.226    0   -13.278     1   
105997       142521         0         0.769   0.716   11    -8.844     0   
105998       145000         0         0.675   0.694    9    -7.689     0   
105999       157335         0         0.642   0.759   11    -7.293     0   

        speechiness  acousticness  instrumentalness  liveness  valence  \
105000       0.1990        0.7510             0.878    0.1740    0.401   
105001       0.1410        0.5700             0.891    0.1270    0.100   
105002       0.0528        0.0361             0.854    0.3800    0.587   
105003       0.1060        0.5760             0.786    0.9180    0.379   
105004       0.3230        0.9470             0.885    0.0933    0.341   
...             ...           ...               ...       ...      ...   
105995       0.0314        0.0238             0.896    0.1530    0.337   
105996       0.1010        0.4930             0.904    0.4460    0.200   
105997       0.0514        0.0161             0.946    0.1520    0.598   
105998       0.0838        0.4450             0.928    0.1090    0.625   
105999       0.0437        0.0285             0.891    0.1940    0.753   

          tempo  time_signature track_genre  
105000   90.076               4       study  
105001   83.030               3       study  
105002  112.509               3       study  
105003   80.014               4       study  
105004   77.168               4       study  
...         ...             ...         ...  
105995  112.018               5       study  
105996  135.046               4       study  
105997  137.995               4       study  
105998  179.960               4       study  
105999  174.036               4       study  

[996 rows x 20 columns]

We have 996 tracks in this genre. it's a bit difficult to see the songs looking at the top and bottom. Let's again look at a sample of 100 rows. 

In [17]:
pd.set_option('display.max_rows', 299)

In [18]:
# 100 random rows from study genre
study.sample(100)

track_id                      artists  \
105453  57bY38pZVvHLS6LCBpzMKQ                 Oscar Hollis   
105861  1O9XqbawlAACCBNT1tg6YM                   Matt Large   
105322  3AA7qG6AgLxmTDAlZVRKaX                Sless Praismo   
105629  0PkaVMW00sCAeiXd5mKIVb                   Odd Shapes   
105592  5nMWNjCkesUlHXjjfBNsAL             Nathan Kawanishi   
105092  72wDz8RDKFvffQJhHucHDy                  Dusty Decks   
105413  4sT5NZD0cHvpB3fcDkrGpD             Bullseye Release   
105859  2vXP0h0jE1vOaPM0zS5Ora             Timothy Infinite   
105894  74f0VaipJke6lIWI5BEfTW  Sarah, the Illstrumentalist   
105885  3ZBJMy7ICSqGfwQivSjqZo                     wavcrush   
105745  3GkknTSc0Ayr54Pq8unwfv                     meadowzz   
105403  3y6KYivzMqOZDscUH2Cn9N                     meadowzz   
105945  5eaVgfh8tX586mdr9utjd7                    Beat Oven   
105022  1D5NGDRg9CFCIm6WXgjtd8                    Beat Oven   
105729  5mhCTQzibAYfmEEAyVAJaJ        Sless Praismo;Frogman   
105300  1bqbXMB4sjKR5QEkZRAm2h             Better Call Soul   
105946  2yo0KzGvj5u2h7JRgX0JJ5                   Justnormal   
105402  68ZIOt33K90kdFTRiHNu9W                        DGHTR   
105120  09be8fMevpQYEOkuD75eC7               Clint Is Quinn   
105434  2fNcvNQROEwslviciOWQ6G                   Blackburri   
105436  5T41z0Wg9mPbPGXvquzaiK                      Ensidya   
105523  0rNoP4LhYjNIjePcKISLyH                    Beat Oven   
105550  31DKCxT3bi5mIbv53jyvBn                    Smartface   
105650  6WLH03pRffbNXkGfUEDmpU                  Joi Casette   
105335  2WsV7i7DLaCJpNWaZ50lzz             Timothy Infinite   
105942  2HV5Ur7jyVA9bwZjc4Fjhp  Sarah, the Illstrumentalist   
105134  2nQqrEnAZgyP3GcT7tMIdz                  Eli Stewart   
105500  07vontPkjk1Wmkar99vNxp                       sourja   
105319  7yEkBD9zVnXHCwhBlM3ffJ                     wavcrush   
105969  27Ji3sj40M4a6roEpL1Box  Sarah, the Illstrumentalist   
105082  37wRNwFhzLSkYx3CrSi9JI         Breakfast For Brunch   
105127  2vwVuPxqckJFtBMIHME8xu              downstreamdream   
105657  4GpJyQ58wvKwllL3CAbfX5                   Justnormal   
105797  2yJNOLlNJcxIq6OupaCDcq                   Matt Large   
105723  18VVHK0tjbcgKWQzM8O5fo           Smartface;King Sis   
105793  2GVHkBOCH81D597d92vYwo                    Beat Oven   
105081  1U6LYBh6HZXaaQu3QJlDLm                        DGHTR   
105970  35vJiRzgPc7Ocb03TjPHyT             Timothy Infinite   
105588  5idpF5W5D1TmaQ1kcvrT4l                    Pat Laine   
105142  7e4xPma019GgBJ9akT42sZ  Sarah, the Illstrumentalist   
105018  1bFNVPttUZyzRxfsJAFkJx             Nathan Kawanishi   
105762  7hbI6QqjMZ5vlcE7A9DcVF                       Almo X   
105625  0TRojooVyTAYS0O0vaNZGW               The Daily Bred   
105457  6FWjkgQnnOV4LvLykqbbCo                  Dusty Decks   
105255  7rH4Wp2iuM6ev0JiKMMor2                   Blackburri   
105975  6NZZQo5RQKR6ZZpIM9oKuZ                        DGHTR   
105465  6kdLIWbpIFJsa38nb2g15n                        Jobii   
105236  1uqbepnCPGf5eZZFKDKvQG                    Smartface   
105112  5fJYaMNnnKiktBETWKYTmD                     meadowzz   
105987  7zNnFntwJVGbOffLoywXUk                     Lomoplay   
105030  0PMlGvzeLrQ27h8oGyfGMV         Breakfast For Brunch   
105217  6ZPMtkVR8cyVCy8vKDxyFM                     beosound   
105689  5XgMwApqJy2vSp9KplToum             Timothy Infinite   
105904  7LMwv7GprGBcAkKMDxda5J                    hope mona   
105516  4iNyEaflNRnwD355Fk23GH                     wavcrush   
105364  4dAbCuNehX0eahmR3uXhpD                       Cerwen   
105343  6pr6WyqZ7y8CszSE334kZw  Sarah, the Illstrumentalist   
105087  3I6fpot5ZS5H5T5YDB67rQ             Timothy Infinite   
105147  57NUJvvaolgDOw4DwumiDN               Wasted Glitter   
105736  2LAidQYi4vzoPl6nkqxyCn         Type.Raw;Loot Fattig   
105144  286NfUkb1H8WjWkDfUESqw                  Sail & Weep   
105848  48YyZ1XTQT9UZrBoVNY568             Timothy Infinite   
105074  0xu4nHvp3S4wVFc685Z9EY               Wasted G

I am not familiar with the artsits, albums, or tracks in this genre. But we can look at the numeric features and see if we notice any trends. 

In [19]:
# Summary stats 
study.describe()

popularity    duration_ms  explicit  danceability      energy  \
count  996.000000     996.000000     996.0    996.000000  996.000000   
mean    26.152610  141592.863454       0.0      0.685067    0.411024   
std     14.121255   33766.767066       0.0      0.099322    0.168141   
min      0.000000   68500.000000       0.0      0.331000    0.028200   
25%     11.000000  121452.250000       0.0      0.629000    0.287000   
50%     28.000000  138666.000000       0.0      0.695000    0.397000   
75%     37.000000  157710.750000       0.0      0.752000    0.519500   
max     55.000000  307246.000000       0.0      0.937000    0.959000   

              key    loudness        mode  speechiness  acousticness  \
count  996.000000  996.000000  996.000000   996.000000    996.000000   
mean     5.427711  -10.846727    0.501004     0.100187      0.530086   
std      3.629000    3.586343    0.500250     0.093228      0.311927   
min      0.000000  -25.150000    0.000000     0.025600      0.000386   
25%      2.000000  -13.248750    0.000000     0.044400      0.256000   
50%      6.000000  -10.319500    1.000000     0.065700      0.570000   
75%      9.000000   -8.012000    1.000000     0.112250      0.811250   
max     11.000000   -3.343000    1.000000     0.743000      0.990000   

       instrumentalness    liveness     valence       tempo  time_signature  
count        996.000000  996.000000  996.000000  996.000000      996.000000  
mean           0.789227    0.161438    0.402181  111.776734        3.942771  
std            0.212534    0.126922    0.213660   35.636203        0.332098  
min            0.000016    0.035400    0.035100   30.322000        1.000000  
25%            0.785000    0.102000    0.233750   82.999750        4.000000  
50%            0.866500    0.114000    0.384500   92.040000        4.000000  
75%            0.909250    0.157000    0.548500  141.995500        4.000000  
max            0.975000    0.918000    0.970000  208.060000        5.000000

With the exception of loudness and instrumentalness, there tends to be a wider range of values. Seeing as these songs are from different genres but deemed to be appropriate for studying we can go ahead and remove these to improve the performance of our model. 

In [20]:
study_index = list(study.index)

In [21]:
# Removing 'study' genre
consolidated_spotify.drop(study_index, inplace=True)

In [22]:
# Sanity check 
consolidated_spotify['track_genre'].unique()

array(['acoustic', 'afrobeat', 'alt-rock', 'alternative', 'ambient',
       'anime', 'black-metal', 'bluegrass', 'blues', 'brazil',
       'breakbeat', 'british', 'cantopop', 'chicago-house', 'children',
       'chill', 'classical', 'club', 'comedy', 'country', 'dance',
       'dancehall', 'death-metal', 'deep-house', 'detroit-techno',
       'disco', 'disney', 'drum-and-bass', 'dub', 'dubstep', 'edm',
       'electro', 'electronic', 'emo', 'folk', 'forro', 'french', 'funk',
       'garage', 'german', 'gospel', 'goth', 'grindcore', 'groove',
       'grunge', 'guitar', 'happy', 'hard-rock', 'hardcore', 'hardstyle',
       'heavy-metal', 'hip-hop', 'honky-tonk', 'house', 'idm', 'indian',
       'indie-pop', 'indie', 'industrial', 'iranian', 'j-dance', 'j-idol',
       'j-pop', 'j-rock', 'jazz', 'k-pop', 'kids', 'latin', 'latino',
       'malay', 'mandopop', 'metal', 'metalcore', 'minimal-techno', 'mpb',
       'new-age', 'opera', 'pagode', 'party', 'piano', 'pop-film', 'pop',
       'pow

Let's move on to 'ambient'. 

## Ambient 

In [23]:
# All the ambient music
ambient = consolidated_spotify.loc[consolidated_spotify['track_genre'] == 'ambient']

In [24]:
# Random 150 rows from this genre
ambient.sample(150)

track_id  \
4694  6p9Frj8pPOCWol6HMFihSs   
4716  5KIOfWtUEKrxDNi4OnIpg2   
4172  4Ks0OfgQpb58nHvQrcaiPi   
4806  1qGI8KYIXtJK9WTMC6nIQM   
4659  4FEcEOyyGbPR1cnYMBTRfO   
4096  6npUdmcsfp9yRa65tF0PUH   
4217  4GEzdXpbeGB97W1C3gH8Hh   
4898  1q6ek8IwZsfPYFrbiuSEOz   
4181  2NZEJxIUnsP18o2aNzeuZW   
4390  6T10XPeC9X5xEaD6tMcK6M   
4106  4TNZTT33erLVg31KU4RJvf   
4206  3lB90ecwKTc3SHHUadsv2E   
4839  3vgWxwRvHJuLXiwtqe6AvY   
4151  0vSGh7m49wOnmPDkv3m9m1   
4436  0dHF8t7OIhgPele5WRUfWG   
4686  5CKDuCZN9qq3As3WDlxK1b   
4956  6tbYNvsJLTfcK0SpJ2LDv3   
4342  21JL0gpeAYgbSUmzhgSPJz   
4724  32zrKU0KPtMBFdQv5LTa9H   
4249  0VLascXUZ2Ze2u5kUD5J1f   
4135  5AbU5I0K9Z7muuvUw4No4n   
4316  7iLVpb0WAgeVRn2wSdmdrd   
4780  4nAvEP0TxzNPiu710Dl5va   
4527  7b2dMRCVDx7OGekIVjCrs1   
4269  06UlQEupfklesVD0OB2N87   
4717  1QGQ7wB1KZfpbbWNFCaVGm   
4770  698afgx5ev4sqwZBTqHpoW   
4077  56oReVXIfUO9xkX7pHmEU0   
4603  4Yf9QLUM8EeHaCS7EnaHnX   
4495  79ykCdN9JJ5DJJ845zAQEY   
4656  5wm0L02zvFgWP6QwIBh16p   
4535  14I9CpiA1CMTdLga8SKAjJ   
4546  43p2e6ACVNhBoV42kEr4J8   
4127  0qf2iUDfPDtrLeJFmOvVn6   
4299  4qC78OGZyxvgX8tJELu6jU   
4994  7gLVKmFdPHGTrrK5QknUnw   
4228  3cBtANnJGopPaRMXCl3mV7   
4581  3HRqtrUKxbPgE1dqpZT1vb   
4346  1vTkNAelvP5V2vbQ5rAbu3   
4932  71dLJx3qHOTQMTvvoE2dmd   
4698  6qFJBjIf3d0XsmpEFmZez9   
4445  1V8aPypxFrIk5TSa8E3qM0   
4066  0wzThSEGQOnrGtFnyxGSMZ   
4352  5F0hIO08012qttN94QlO7k   
4120  13Z3uZXTFpbLBZ8R1rTUKO   
4270  0OF6WSdeVmYEGBZlxvwvLq   
4162  21lsdnxagzViwbRqNY0tCK   
4732  7tUQIhueKYGnKThIseP69M   
4095  6yvAfQ4xM01Sq00DZ54sDl   
4336  6dxs9Sv6Ic3oQ3tl6Hozi2   
4198  1MPafzHDGPTx5T08I5cnr4   
4216  5KNZME6ZAXNA4Ce6iOedU5   
4618  1WOxGNutRwrbik8BA5f2AY   
4708  3Yagxg6rtkwN01AYzA862i   
4373  3Hv2XS0W9g4FAGzXCOYEnn   
4388  2DrarKE3JnOXED3kbTuLtY   
4112  1g0xyCMUBKUaMDCk00Hhmk   
4412  7vbCfkJf89i4s745KbELgr   
4768  58PkDDbzS5ptoXaqqBuuoX   
4376  0rFdP3ZfwsUajBLUuex0Kb   
4687  15m36xE78aYbp6Z9W7TBXv   
4584  4vwtWzCqbnSrmKE6lZyUL3   
4750  1XmdB9Kha5e5H4PfTSmewn   
4598  6d3RIvHfVkoOtW1WHXmbX3   
4824  10N3wnHQRBRcaiXjslxmZl   
4507  714swlrTc262NA0j13C7gY   
4729  0Gfi0NProXzPEQbgBaz48D   
4076  1gj9poFl2mLKkguqWo5Y1i   
4226  35U312XkyaOTEbaftUt1K6   
4832  5tXCyx9RaGX7RHcWDWcm88   
4111  5eG8QuQKOBvDeCyoOeMkhT   
4562  5oiNJJdKqJ5Vfwion1dR1n   
4605  7fgqEIEV6kpswoBsgru92Z   
4798  5NJiqDbIe4bb0BIkQzkJu3   
4853  7I8v42wxDhfojsq3APrXU0   
4809  45m8T6QIgYoCZnUNxYeyml   
4463  6jPob3XDIVyNwPKUv2GwBf   
4595  1t4EDMSxXLQhOiutcMN6VQ   
4121  7gH3Oc5AaQs54XtEoLovox   
4180  03b9EWz24mTfDvnrE8G41I   
4759  2YlLv22mNqbG3logspCLZk   
4399  6EXO7CcK6z7A5CPjicb4H5   
4047  5nYs8umrPbDr3rXkCS74AX   
4485  3OpfrN3L1WDBXUYC2z2b8p   
4470  2GpTaKi3KDCQnyMFgwxzXk   
4963  5JZq6B2jtxGVvmausM9Xed   
4499  1ZJ2XcPB5r4yK08gvJbZdW   
4804  4fF4M4RTXEHpTPMa7c42RX   
4944  3bCmDqflFBHijgJfvtqev5   
4008  2kfGoV9a5dbSKCNmUWH2ZF   
4741  3rLe832ZEZfpBmkgXGSrtX   
4911  7aK1caHaXe63BtAoIFvN2h   
4703  1zl2OD6MgQoXlQJbXnXfJL   
4261  2dHnIe0JsqshUnyzcMMOhW   
4530  3JlVI5OZuBFhXQ9GUFdUxN   
4356  2fgpYCWp31JJlplZ7MRANG   
4185  2k2MprlqkmDMKKOOWGREzW   
4504  4azoV50K4s4eBkGRHSVWYV   
4692  2WkuESwWyJ1txlbXnDIwPm   
4764  4g0QeSC67nz3qwHgyx038U   
4266  2RaUxxkie3gGE385mTEODv   
4491  29S6PLRccSCyKozzwIik0J   
4321  1GfdC8GouQqL0RKKzgIhXw   
4959  3skG5verLx4Tq6BHZ3Hv64   
4904  0WBNZ9HEz6QqiDuZ3BCehP   
4561  1QSfgGcP3Rmdo2EdW1s4ya   
4658  3JggEa7I7tBPQhiOfexL6U   
4354  2unMCmT3nl0VBRFI3J0xIv   
4749  3RzCcZ05bVMHfsebF5PHmw   
4378  72SrSfTI6p9HCBuVKmCNu9   
4890  3TuN88sSwjNy1RAUDiKTm1   
4486  1QGlEQJkixOXsbijKqNTDB   
4921  7eTDwpm5XKKssoMhMbo67y   
4176  1ulB5740qjdSila4Vjnb7Z   
4459  47aMbtetHwI7tLR9Z7iec7   
4558  6f4cBWeUGOtO7WFxxt35j3   
4465  7eL7yfCL9hs0mAW6B5Rfmg   
4778  56ECskEyc7XqOCyRPruZ37   
4993  3HHOfa5QV20fxF12JPgLsj   
4863  16D9AzevsZAh4GfQEYIo7I   
4431  5shiLLxKQ4U7XLELZeXzxy   
4298  3Oi0avSYPw3ZAkemFGwr6U   
4849  5v3TSHYm8BzbON2u6QBEG7   
4979  0vp09QgBhShrQOL2w7P8k4   
4521  0bnSSnjn83WSf2

In [25]:
# Summary stats 
ambient.describe()

popularity   duration_ms    explicit  danceability      energy  \
count  937.000000  9.370000e+02  937.000000    937.000000  937.000000   
mean    44.817503  2.361337e+05    0.005336      0.367037    0.235512   
std     17.409528  1.128798e+05    0.072893      0.164229    0.210397   
min      0.000000  4.190400e+04    0.000000      0.000000    0.001440   
25%     40.000000  1.600000e+05    0.000000      0.230000    0.070900   
50%     50.000000  2.182000e+05    0.000000      0.361000    0.176000   
75%     55.000000  2.895480e+05    0.000000      0.487000    0.348000   
max     84.000000  1.041520e+06    1.000000      0.853000    0.974000   

              key    loudness        mode  speechiness  acousticness  \
count  937.000000  937.000000  937.000000   937.000000    937.000000   
mean     4.955176  -18.679177    0.616862     0.041793      0.776710   
std      3.444945    7.957414    0.486411     0.030389      0.301903   
min      0.000000  -41.808000    0.000000     0.000000      0.000002   
25%      2.000000  -24.141000    0.000000     0.033500      0.722000   
50%      5.000000  -18.199000    1.000000     0.037600      0.921000   
75%      8.000000  -12.016000    1.000000     0.044600      0.979000   
max     11.000000   -4.431000    1.000000     0.761000      0.996000   

       instrumentalness    liveness     valence       tempo  time_signature  
count        937.000000  937.000000  937.000000  937.000000      937.000000  
mean           0.676701    0.128756    0.166526  111.290487        3.648879  
std            0.361738    0.096965    0.143450   32.057238        0.775893  
min            0.000000    0.034500    0.000000    0.000000        0.000000  
25%            0.505000    0.091600    0.056900   84.909000        3.000000  
50%            0.877000    0.105000    0.125000  111.995000        4.000000  
75%            0.924000    0.119000    0.230000  132.614000        4.000000  
max            0.986000    0.975000    0.962000  213.848000        5.000000

The wikipedia description of ambient music is a genre of music that emphasizes tone and atmosthere over traditional musical structure of rhythm. It is a loosely defined musical genre that incorporatees elements of a number of different styles - including jazz, electornic music, new age, modern classical music and even noise. 

Just by checking a few artists, we can confirm that definition. Many of these artists span many different genres including but not limited to classical, indie, alternative, and electronica. Ambient music is a large, legitimate genre in its own right, but the mix of styles could make it harder for our model to perform. Using our fairly basic attributes, it might be difficult for the model to analyze this genre. For example, instrumentalness can vary quite a bit. However, energy, acousticness and liveliness have values that lie within a small range showing possible distinctive features of the genre. 

Let's leave this genre for now, particularly for our Principal Component Analysis. It's possiible we will be able to combine this with something else in our data. We can also try modeling with this genre and without it to see if this should be seen as more of a sub-genre in this context. We can come back to this genre later. 

Let's move on to `sad`

## Sad & Happy

In [26]:
# All songs under 'sad' genre 
sad = consolidated_spotify.loc[consolidated_spotify['track_genre'] == 'sad']

In [27]:
# 150 random rows from the 'sad' genre
sad.sample(150)

track_id  \
94974  60QvDCrjIu6DmewWRxKKRa   
94019  5bryfrnYaxlofC6MAG6vde   
94435  4BUEgtfQSZkHbjjztrKtlv   
94220  0CL5CGhcKaJLakMZNSRLu9   
94400  1PJsUeqqU2H2iVCZ7ZBGnU   
94678  2pLPUx9oQfg8L5sAEPDGBp   
94100  7xcTzEFNkkfO5a60T93tdk   
94410  3ydusC725etWQ7QlAjmIi4   
94034  56H3K8a3gbWeX0cg4GjSzC   
94647  50HJE4g0GR62ID38kzRMGo   
94817  7bckng9psD3hFOCiIi0qqT   
94288  4yIHTPSFg5Ma41IJNE0GUL   
94849  5dVFF9IrNcPoBKAviJcKC1   
94234  0RCgQNZ5SAjPukI3OXquFN   
94352  2EZy2nbJyUpIfPdbTyZbRO   
94604  2ugb7Vrnvw8wwM6DpDNiPl   
94017  1xhiKL4UgfkAfnCUg6HEpA   
94358  0TGIw5lMfCjMhFZN7z43uk   
94800  1uuI4ovQALuaq9srVZyTDy   
94847  0yldWDxxRIsO0BqPlVDd8H   
94840  3MrZcAFdNmtPVbob9QYvs4   
94808  31qlV50cTaXoZGt0G3AQfo   
94457  6sx0LC4zyYKtFsW3fmDfsk   
94900  4Jmkp9Rpkl4FiGqX8O0IpP   
94901  3icSLmnxMhqKwjVoHzMHyj   
94330  7CHMY5TN1e8oqDnRwwJR61   
94958  6qsVVFBn0CuPrEgx9u4QzK   
94171  1TYF5FyxFXC6bM2VOuNWhV   
94114  3WaEyZXCEKX8TxmKciVNHm   
94139  1kFlDnWmBTvCBxxeGDpIdp   
94579  6bWJ3kPjskTCS2vzPwl9tY   
94778  2Zql6aBjLdgYTwhIdjXDGg   
94749  72542euQT9wECESPCIKSXS   
94275  2EL7PmNeZuFSWPYbnMS0Sn   
94416  6KA2YcsQZlQD8f8uvKjPXw   
94123  0yI1LOVVDIN5QgnMdUo3Ym   
94037  4E4DQVf8GNBvvOaLfvIcvZ   
94283  6SdW1PtDQZ06dumP6KUjJT   
94866  3zefoLCDoPU5PJK6g1EHoF   
94822  7a5TsU5Nw3VPrR1HH7mHJN   
94299  2ZQLEuaDw9rxJ5u5YR0KEw   
94502  5zxF5OGQKFQTPZoKUaY35q   
94231  4vsLtmWxiRMhzW5ZkGw7mX   
94140  3Qx1CtexENLotadyrPxGsY   
94112  1Q7HO1L7KO46VlMH8C3Abh   
94831  19coiw9dDWBNHtQhdKL1VC   
94948  69XgyxmE6zl5SmHf2NvjqU   
94490  34eD3YuDcGztZE8yc1MGB4   
94425  3p8iG863bPsafyReUXaqXO   
94039  33LizV83s7z1gwtmLiKlmv   
94869  2aoNxQM2bTWqF2uC0e8oV0   
94322  7sQhhSlMui3JOZ4s2j4IjM   
94888  0WMDVSpzpRBuRqMWbbbYVI   
94966  7xK0LBgSUW1RBUHrzifjP3   
94021  58wyJLv6yH1La9NIZPl3ne   
94052  7cPQWcczApOEcmeFwUG9xj   
94989  7gdGFqMtL5XFHRP36N2Qdb   
94031  01UeBfllW39eaKdiCw556R   
94106  4HlwrvZ1Onf2Vh8PJtIvtE   
94544  0BZagFWMAL33p7y6HsL3jT   
94323  6hbK9zJUAlB5PYL15SzKtB   
94104  1c03f9YZmO792D7hRjZMYi   
94771  68gRQeI1JL6ELZ6juJh7N4   
94828  2GCAgfxHMJYMqWYIVfNDum   
94272  2ZLoprAEuvfs30Y4InKlaq   
94362  6MjDnmBgbCu2UfDK0DRczh   
94306  3awnZSyvKdi7rBLttHk36K   
94646  1ux2H51GHPjSntjRvxU2gN   
94660  29c5U8gKc38Ck9j10g4A8K   
94766  0SzsVQB4gK7KRHWycVeFeZ   
94463  0UX0TejWosIkJLtXeXJ1F6   
94475  0icgGg4BN47ILHEhOSaFgh   
94141  17oY2Rh5qQH6mhyakZmpxk   
94663  3TRwFfzKktnElZbdmcrmem   
94645  0iEIyCagJYU1NU5ukOEnAL   
94094  50OPOdJbRx9iCJXGultmUf   
94807  0L1qOe6ZGo8NZeU3yLclZP   
94951  79DvIaSQZlen3UTJrjNwut   
94179  0k2bhZ9oM3boyUWaV3MfWk   
94496  7LBwwQJZChGm4904z8ZXgU   
94657  1hQaFnjMPNskSo05giqiVR   
94803  2XjJeWTDbI0CtGUsMHlTXy   
94918  5NICLsAN2wxcFQEdriIV1s   
94081  2rpW6YlfARNrITNWJbqCgM   
94452  5OdIMoELVs6AE5KifEOWWw   
94058  2L2mNkEWo818IE4fKbnP0O   
94229  7dTTlQbEBPnU1qA4YYVLqj   
94931  2AkvoevjI6ABNHFTbCIWzT   
94108  2NYPx7YjNrHteC2RxzTzTZ   
94030  0WSDEpEZZDffYfmK8Ssuxc   
94726  3s4WVKaSU112SfRBjvyXKp   
94374  2E5xToEkMtmegq9r7GZYgV   
94622  4njf8PZi95E9ISlQVXXPyl   
94740  0I1EHCAxMY9n2KTjiZik54   
94874  2Y772St4hdHDHvRuIkVwnv   
94843  3wHLn7OTV6d4Ai3oU2d1my   
94091  4EvVZsbEh519mwqxgOCx6v   
94991  5yZsE3TKmxdRVrSc8TCqfx   
94924  4YnKb0qfI6j56S0UiaewIC   
94076  0o2Jngx6YVivDjE6RqqFlF   
94001  4ZLzoOkj0MPWrTLvooIuaa   
94985  3M93msz01FcdgUUG9Zl4WV   
94652  0x2HtOh21kUsh02xmnWdw5   
94936  5Z1KaSy7mMOThbYBBjaBR5   
94894  1DogGoIMkFf8jsrs3AlTUQ   
94846  5612zrK8d3HhPCobx2Kuci   
94510  5VPSuEwoiGMvUvczjbjJSE   
94309  7HFGvz6pWt8RW3sA6Orl3k   
94409  3HD9BE49dhRLZxFIzoRJg0   
94983  5nMTWDeBtWEaeJbejhu7Jp   
94576  6ewXjgNwJy2Dhc4mvVMbfj   
94999  5Okrz6AYoHlzGRWLEeSrrX   
94268  3AlVajn3U2MEuTeh4Z26vK   
94421  1UubQzL9nNulJW5DHqMKVO   
94233  7G4uppQl36EqokV8TyzP8Q   
94447  3OQV3UfK15wNcinIxdBvoA   
94406  3pVerm3v4HIO9mlhDIqxpa   
94796  3KH53JvvNZD2KpvlT1ZeJZ   
94363  3ZSEV608VISnC8opWYmR2O   
94628  69FiPptnZd9ovg2tEJgBkN   
94095  0rIrIUpcd1FFRSEwQGk1C

In [28]:
# summary stats to see if we see any patterns 
sad.describe()

popularity    duration_ms    explicit  danceability      energy  \
count  536.000000     536.000000  536.000000    536.000000  536.000000   
mean    51.807836  152524.983209    0.501866      0.703657    0.478351   
std     10.831005   36723.805015    0.500464      0.126381    0.164225   
min      0.000000   57877.000000    0.000000      0.328000    0.033700   
25%     47.000000  127741.500000    0.000000      0.624000    0.354750   
50%     53.000000  150000.000000    1.000000      0.719000    0.481000   
75%     58.000000  175533.000000    1.000000      0.798000    0.602000   
max     83.000000  314500.000000    1.000000      0.980000    0.906000   

              key    loudness        mode  speechiness  acousticness  \
count  536.000000  536.000000  536.000000   536.000000    536.000000   
mean     5.402985   -9.775991    0.617537     0.131899      0.435505   
std      3.667222    3.758678    0.486443     0.119375      0.305249   
min      0.000000  -23.236000    0.000000     0.028100      0.000496   
25%      2.000000  -11.820500    0.000000     0.046500      0.139250   
50%      6.000000   -9.036000    1.000000     0.085100      0.409500   
75%      8.000000   -6.997500    1.000000     0.176000      0.708000   
max     11.000000   -2.325000    1.000000     0.891000      0.991000   

       instrumentalness    liveness     valence       tempo  time_signature  
count        536.000000  536.000000  536.000000  536.000000      536.000000  
mean           0.070964    0.163116    0.437461  119.812062        3.983209  
std            0.211245    0.113619    0.207810   31.396476        0.283004  
min            0.000000    0.044500    0.038700   58.968000        1.000000  
25%            0.000000    0.102000    0.273750   92.996750        4.000000  
50%            0.000000    0.120500    0.434000  118.711000        4.000000  
75%            0.000407    0.175000    0.590750  142.289000        4.000000  
max            0.973000    0.851000    0.970000  220.039000        5.000000

We do see that 75% of the data has a valence score of 0.59 and below. 

Based on the tracks' content, and some metrics like danceability and energy we can assume that the title of the genre refers to the lyrics or sentiment of the songs not any musical style. This in itself is a good reason to drop this genre seeing that we do not have a metric that quantifies the meaning of lyrics. 

In [29]:
# All indexes of 'sad' genre songs 
sad_index = list(sad.index)

In [30]:
# Using 'sad' indexes to drop them from the DataFrame
consolidated_spotify.drop(sad_index, inplace=True)

In [31]:
# Sanity check 
consolidated_spotify['track_genre'].unique()

array(['acoustic', 'afrobeat', 'alt-rock', 'alternative', 'ambient',
       'anime', 'black-metal', 'bluegrass', 'blues', 'brazil',
       'breakbeat', 'british', 'cantopop', 'chicago-house', 'children',
       'chill', 'classical', 'club', 'comedy', 'country', 'dance',
       'dancehall', 'death-metal', 'deep-house', 'detroit-techno',
       'disco', 'disney', 'drum-and-bass', 'dub', 'dubstep', 'edm',
       'electro', 'electronic', 'emo', 'folk', 'forro', 'french', 'funk',
       'garage', 'german', 'gospel', 'goth', 'grindcore', 'groove',
       'grunge', 'guitar', 'happy', 'hard-rock', 'hardcore', 'hardstyle',
       'heavy-metal', 'hip-hop', 'honky-tonk', 'house', 'idm', 'indian',
       'indie-pop', 'indie', 'industrial', 'iranian', 'j-dance', 'j-idol',
       'j-pop', 'j-rock', 'jazz', 'k-pop', 'kids', 'latin', 'latino',
       'malay', 'mandopop', 'metal', 'metalcore', 'minimal-techno', 'mpb',
       'new-age', 'opera', 'pagode', 'party', 'piano', 'pop-film', 'pop',
       'pow

Based on `sad`, and the fact that we have the valence metric, we can also go ahead and drop `happy`.

In [32]:
# Grabbing all songs with 'happy' genre
happy = consolidated_spotify.loc[consolidated_spotify['track_genre'] == 'happy']

In [33]:
# All of their indexes 
happy_index = happy.index

In [34]:
# Using their indexes to drop them from the DataFrame
consolidated_spotify.drop(happy_index, inplace=True)

In [35]:
# Sanity check 
consolidated_spotify['track_genre'].unique()

array(['acoustic', 'afrobeat', 'alt-rock', 'alternative', 'ambient',
       'anime', 'black-metal', 'bluegrass', 'blues', 'brazil',
       'breakbeat', 'british', 'cantopop', 'chicago-house', 'children',
       'chill', 'classical', 'club', 'comedy', 'country', 'dance',
       'dancehall', 'death-metal', 'deep-house', 'detroit-techno',
       'disco', 'disney', 'drum-and-bass', 'dub', 'dubstep', 'edm',
       'electro', 'electronic', 'emo', 'folk', 'forro', 'french', 'funk',
       'garage', 'german', 'gospel', 'goth', 'grindcore', 'groove',
       'grunge', 'guitar', 'hard-rock', 'hardcore', 'hardstyle',
       'heavy-metal', 'hip-hop', 'honky-tonk', 'house', 'idm', 'indian',
       'indie-pop', 'indie', 'industrial', 'iranian', 'j-dance', 'j-idol',
       'j-pop', 'j-rock', 'jazz', 'k-pop', 'kids', 'latin', 'latino',
       'malay', 'mandopop', 'metal', 'metalcore', 'minimal-techno', 'mpb',
       'new-age', 'opera', 'pagode', 'party', 'piano', 'pop-film', 'pop',
       'power-pop', 

Now that we've covered loosely defined genres, let's move on to the languages. 

## Languages 

We have several different genres titled with a language. Due to our limited attributes, let's look through these genres to make sure that we don't have overlapping genres simply sung in different langauges (i.e. french hip-hop, french indie, french pop, etc.). 

Our languages/countries are: 
- brazil
- british
- french
- german
- iranian
- latin
- latino 
- spanish 
- swedish
- turkish 

And in addition to the above let's check out `world music`. 

### Brazil

This could either be popular brazilian styles like Samba, Forro, Bossa Nova, etc. or brazilian "versions" of the genres we already have present in the dataset. 

In [36]:
# All the songs that have 'brazil' as their genre
brazil = consolidated_spotify.loc[consolidated_spotify['track_genre'] == 'brazil']

In [37]:
# A sample of 150 rows to randomly assses content
brazil.sample(150)

track_id  \
9880  2igDtZlgWm3ZxhV7CHv6pI   
9881  7x4bh40tbYvpN5Gp7fMyts   
9526  5mblm2qeUbMbsDNsIE5OQi   
9437  6M1mIjeq2kNvT8F2dHNugk   
9767  1muDihGGBhRaAfIDUA93z9   
9796  6WmlPRkJH68P7tmkFuQe3z   
9616  6ijPGzgs3jqHokPhkhPpMh   
9858  3yz4ypggDnJ4N3z3YG6pn2   
9019  6OSs0dmcVqVdeWOmsaA5g1   
9021  2OmZjKpXjbDWXEsw1JEy1x   
9983  7yLysy4GtqecVYpC773q7g   
9330  3qzgGSGtGAcXGZJ7UG68hr   
9907  0mOTIdOUMEhiY9OQ6Rz2Wg   
9993  5aulWAWeV1uSJUJmDGmrw7   
9027  4Lfod3ZkXkJBZZhw2cHKCk   
9960  0c8GzzMZ2ua7ybYhZvp9wX   
9504  3xra5OMlRc4MQgD8V0aIxq   
9705  6ueWea9Fl5OkarsQf4Fn14   
9857  1eJCpFxPd8vKfQgNSvVtLP   
9529  5f1aVfi9nYxMHpcA6h0wOK   
9807  4V1GX6kcU1ehDP1d1GopGb   
9711  1eK1ZyBymaRWFcZBjY8dmL   
9831  4Pi0zgPzwi4MLjoLsPwfEm   
9182  3Ob98mJElVwt9PwhuFqtIP   
9474  4liaba6RDgz3sXEcTjl4Dw   
9938  2NQ9c78cqdeKegSDRSttcj   
9714  37GYyTGRnPcP01Ff7ziOoi   
9883  2a8V54o2V6QdgofUotLu9f   
9035  0ZBYKXS2sFtFOiJNt97mPP   
9210  3B4NyexkxpgmSEikcZJXcZ   
9387  4D75ixy2I4ft75Ay25fKyP   
9666  7DCVTLzUHbfJ2gOUHOCyYu   
9404  6keP0AfCE5v4CTI1E5nrLv   
9190  0eda3exx6ThiQBVQ5q52iP   
9830  0iRUshq6MVKEPuwJfshhM9   
9453  1Iz2RJFAbcyAwB9sOaXvLp   
9623  4B1A052R2n4HGE9JeieF5K   
9106  49PonczSLhra0DxsSdSxq8   
9090  3Dm6aXLhzureDjBZIqFYhK   
9724  3gi7gH8JbG0QdBM4Suf3lr   
9128  7abcm7Q3r2MKCTttMjIZo4   
9109  2CNSHg6E1xchwIuKuQLHtD   
9652  1HQoGpPJ8ye2BEavUqdnF9   
9052  0bJzyiMBy2hpqhgf4RN20J   
9510  3myP5D9V78ayZyEb7cUGRw   
9678  5kNCog1amNsI8z2iD5634p   
9280  1dIah1UaJi3n5pwOaOd5QV   
9924  07CgSl5vWjcXMX4zYkbrLX   
9751  4RdLeGrUDuPsUrQtG0lHEk   
9184  44ia764XUgyUzPVHDALBYt   
9533  6RJ1V6qO8J8Spxo5jsC4wO   
9619  3L6EOy5OG9JetMXAbEJQu9   
9847  469w8UmJYEGGMTNafmscjz   
9476  3kgCFhVnBZbmdBx5LiGfsF   
9065  6ULtPMzIdsayMraZMs2KBV   
9099  1YlCyC5mesTP4h2009JkCd   
9408  4XuEiko0gQBN73HsRTNhkP   
9614  6F1kCUTjvdALqB6p5VkqwR   
9873  0FiYM9wfGccAJn8z4S5hxB   
9436  5diYEOacejP4yRkWlYU3Cu   
9130  5h7e9eHY6RnP617D9zzzKb   
9860  23l6b9hLwkNJGTZJoJPRAY   
9169  3ZbbTKiMX1p7S1hUXtb9SO   
9297  73iM4TuuYXDB6NKIy0zZYn   
9230  2JSxHhnDHsz0d6v22amg7y   
9214  6eQjwvQmfY4vBs4c9IKiuW   
9459  4A4POOAdOR8Jq644tmaMHQ   
9622  1EUm3fTMaakB3KdffkvCZk   
9783  7uNbu1BgkWXTA1nZRX9fZb   
9203  7fjogJNNk2hCWtAKoTRIPr   
9966  574pBc93Hwmzows86H5fSx   
9525  1YffYMKpricZh02Sh5nTMK   
9191  0g8L4DJyzH7CG0nPqGtNlR   
9000  6GomT970rCOkKAyyrwJeZi   
9281  24Jh5ZcRVUQCFnWVO1j8DY   
9044  4MneVvYOxU1LaySFx8n98p   
9251  7F7lizHkjMeo70dYOU3r4x   
9779  0v9mFL6DK8xiyJWNPYjXca   
9686  43gu1CX4K2wdXWVvufsZv7   
9074  0rYFCbRb25YpQ7fx2Gnr6v   
9591  6jZeCjc01WUhtMwEmQLlhg   
9132  6G45Qr6vVxnHxX06qcGEp2   
9788  07LseRQRWlhpFJZYfxD6ah   
9153  4horm0SEWe8kMxFjdf7Aab   
9306  7sJlB1DvS2Co4TG5QqfeCH   
9129  50RxCp8dmcpd5P66b95u68   
9329  7dnY16FGz0MnuW3BRipub3   
9688  4KzSlU2JAYba5hhpIkkfGo   
9197  6op6oqy6YayTHBwq9eikH6   
9598  68mW6JloRbnPziPftal3Nm   
9958  1ZWj2sTMhpZaOa8F1KAPVj   
9834  7nE91jTdniDuc9BS2Haoyc   
9505  09Y4QnnWgQXkT07iLoEOKS   
9273  2anDnmAKIauJ4of6K1BoGx   
9567  4nEFEGZQ5odJGjb7xm2kDh   
9953  6C8skGKGztGq74nzCrq90G   
9468  52UwxcBnQeWh4HeMngkuVs   
9361  0DHpr7eLBrRZvlXa4sNwhd   
9290  7zPds0jMIhjayTMhALd8FZ   
9327  4lAvnZlPMjcRMJ8elV8THX   
9181  6hxdEJHw5qPaA4r0GFGkhj   
9058  5MYKqp6J8rstffAQBCE2nl   
9356  3OIzbnBlkSnfCNb4Xh1k5k   
9669  4ErhOgJwZOTI59NXimTmSN   
9024  0Y9nHU2Fhl76LHbRRmNQ6X   
9676  4qeKW6iHErETwNb4sgdYkZ   
9774  3GOkkTrUcv0ICSznFW2B8J   
9787  0Nf3jUIjWFSZfyDbzHnQsk   
9286  5ShkPfQmBWA1BT2uqu8S7Q   
9351  1hB4340DP1nnBmfWn37uTN   
9239  5I8DslpztOwOpwERbGGDTC   
9433  2ARVjHnMtYgxds5zkuVXfZ   
9089  3iw959smgKf8scwjRCRjgZ   
9398  6IzjGEM0s8Q3avi1DHc5Fc   
9572  5Bo0ayLXkp6ttsjCWJvOHJ   
9695  3SdJZFJWYHymvxUrMPAXdQ   
9060  07WWwzmVHoEMgcbuDo8P0N   
9341  2KezizwgtM5kC8ZQxcBf0T   
9922  2ocyydRHIscl3eEDxBTgAA   
9484  1J17nBsbh9HWVvtJFsAPYb   
9307  7dl6sU2RWIcTVOSa3E2VBW   
9121  2DXAv8MRfb1uoJTF827CK0   
9131  4N8rEj0nz6oVhyv3aggHfP   
9004  1XqWGzOdE6Q6FgFwAg0t9R   
9708  4TOY0nJD6T0xhT

Listening to the tracks confirmed by assumption that these are songs of many genres sung in portuguese. Once again since we don't have a metric for lyrics or languages, it's safe to remove this genre. 

In [38]:
# Grabbing indexes of brazilian songs and removing them from the DataFrame 
brazil_index = brazil.index 
consolidated_spotify.drop(brazil_index, inplace=True)

It's safe to assume that we can also remove the remaining langauge columns (i.e. British, French, German, Spanish, Swedish, Indian, Turkish, etc.). 

In [39]:
# A list of the language genres we wish to remove
language_columns = ['british', 'french', 'german', 'spanish', 'turkish', 'indian', 'swedish', 'latin', 'latino', 'iranian']

In [40]:
# Getting all the songs belonging to the genres above 
language_genres = consolidated_spotify[consolidated_spotify['track_genre'].isin(language_columns)]

In [41]:
# Sanity check 
language_genres

track_id           artists  \
11000   0DuWDLjriRPjDRoPgaCslY             Adele   
11001   3di5hcvxxciiqwMH1jarhY             Adele   
11002   6DEMMeWXfmFAXgDUMMzeg6             Adele   
11003   0gplL1WMoJ6iYaPgMCL0gX             Adele   
11004   3bNv3VuUOKgrf5hu3YcuRo             Adele   
...                        ...               ...   
112995  7fo4ScxpiBnESWu8NRMyzx        Ebru Yaşar   
112996  2iSS96WasflS5EbkXytwDM   Can Bonomo;Ceza   
112997  5MaKcM9OtEnbu8YAZbPtfS      Melike Şahin   
112998  4yiW7dLPFKyMlalsB8UMSc      Hayko Cepkin   
112999  3wfEDxGgqZnuOUerhVAQ8S  Grup Gündoğarken   

                          album_name  \
11000                             25   
11001                             21   
11002                             25   
11003                     Easy On Me   
11004                             21   
...                              ...   
112995               Haddinden Fazla   
112996                  Kâinat Sustu   
112997  Merhem: İlk Konserler (Live)   
112998              Aşkın Izdırabını   
112999         Ankara'dan Abim Geldi   

                                               track_name  popularity  \
11000                                    Love In The Dark          78   
11001                                Set Fire to the Rain          80   
11002                                           All I Ask          74   
11003                                          Easy On Me          85   
11004                                    Someone Like You          81   
...                                                   ...         ...   
112995                                      Öldüm Sahiden          37   
112996                                        Terslik Var          37   
112997  Uykumun Boynunu Bükme (Live @ Bostancı Gösteri...          37   
112998                               Geç Kaldım - Akustik          37   
112999                                       Gibi Gibiyim          37   

        duration_ms  explicit  danceability  energy  key  loudness  mode  \
11000        285935         0         0.331   0.341    9    -6.057     0   
11001        242973         0         0.603   0.670    2    -3.882     0   
11002        271800         0         0.591   0.280    4    -5.494     1   
11003        224694         0         0.604   0.366    5    -7.519     1   
11004        285240         0         0.556   0.319    9    -8.251     1   
...             ...       ...           ...     ...  ...       ...   ...   
112995       249996         0         0.277   0.674    7    -5.128     0   
112996       200776         0         0.676   0.799    7    -6.933     0   
112997       280469         0         0.488   0.672    9   -11.690     0   
112998       206680         0         0.425   0.405    7   -12.428     0   
112999       303670         0         0.691   0.406    2   -18.681     1   

        speechiness  acousticness  instrumentalness  liveness  valence  \
11000        0.0309       0.52800          0.000000    0.1090    0.152   
11001        0.0249       0.00408          0.000002    0.1120    0.446   
11002        0.0283       0.88900          0.000000    0.1240    0.348   
11003        0.0282       0.57800          0.000000    0.1330    0.130   
11004        0.0281       0.89300          0.000000    0.0996    0.294   
...             ...           ...               ...       ...      ...   
112995       0.0458       0.22400          0.000000    0.0943    0.356   
112996       0.0412       0.08970          0.000000    0.1330    0.568   
112997       0.0365       0.02190          0.000035    0.7120    0.463   
112998       0.0299       0.33300          0.000025    0.1360    0.382   
112999       0.0394       0.00747          0.000000    0.3580    0.431   

          tempo  time_signature track_genre  
11000   109.821               4     british  
11001   107.993               4     british  
11002   141.916               4     british  
11003   141.981               4     british  
11004  

In [42]:
# Turning their indexes into a list 
language_index = list(language_genres.index)

In [43]:
# Dropping genres from DataFrame 
consolidated_spotify.drop(language_index, inplace=True)

In [44]:
# Sanity check 
consolidated_spotify['track_genre'].unique()

array(['acoustic', 'afrobeat', 'alt-rock', 'alternative', 'ambient',
       'anime', 'black-metal', 'bluegrass', 'blues', 'breakbeat',
       'cantopop', 'chicago-house', 'children', 'chill', 'classical',
       'club', 'comedy', 'country', 'dance', 'dancehall', 'death-metal',
       'deep-house', 'detroit-techno', 'disco', 'disney', 'drum-and-bass',
       'dub', 'dubstep', 'edm', 'electro', 'electronic', 'emo', 'folk',
       'forro', 'funk', 'garage', 'gospel', 'goth', 'grindcore', 'groove',
       'grunge', 'guitar', 'hard-rock', 'hardcore', 'hardstyle',
       'heavy-metal', 'hip-hop', 'honky-tonk', 'house', 'idm',
       'indie-pop', 'indie', 'industrial', 'j-dance', 'j-idol', 'j-pop',
       'j-rock', 'jazz', 'k-pop', 'kids', 'malay', 'mandopop', 'metal',
       'metalcore', 'minimal-techno', 'mpb', 'new-age', 'opera', 'pagode',
       'party', 'piano', 'pop-film', 'pop', 'power-pop',
       'progressive-house', 'psych-rock', 'punk-rock', 'punk', 'r-n-b',
       'reggae', 'regga

Let's see where we are now in terms of total songs in our dataset. 

In [45]:
# Checking the size of our DataFrame
consolidated_spotify.shape

(69844, 20)

### Other Remaining Genres to Remove

We now have 72,939 songs. That's a significant loss. However, after removing the following genres we can start to combine some of the remaining genres together based on their umbrella genre: 
- children
- chill
- comedy 
- disney
- kids
- romance
- party
- piano 

These genres are either grouped based on their lyrics or they contain tracks in overlapping genres. 

In [46]:
# list of genre names we want to remove
remaining_genre_cols = ['children', 'chill', 'comedy', 'disney', 'kids', 'romance', 'party', 'piano']

In [47]:
# Finding all the songs that are in the above genres
remaining_genres = consolidated_spotify[consolidated_spotify['track_genre'].isin(remaining_genre_cols)]

In [48]:
# Grabbing their index
remaining_indexes = list(remaining_genres.index)

In [49]:
# Removing themfrom the DataFrame
consolidated_spotify.drop(remaining_indexes, inplace=True)

In [50]:
# Sanity check 
consolidated_spotify['track_genre'].unique()

array(['acoustic', 'afrobeat', 'alt-rock', 'alternative', 'ambient',
       'anime', 'black-metal', 'bluegrass', 'blues', 'breakbeat',
       'cantopop', 'chicago-house', 'classical', 'club', 'country',
       'dance', 'dancehall', 'death-metal', 'deep-house',
       'detroit-techno', 'disco', 'drum-and-bass', 'dub', 'dubstep',
       'edm', 'electro', 'electronic', 'emo', 'folk', 'forro', 'funk',
       'garage', 'gospel', 'goth', 'grindcore', 'groove', 'grunge',
       'guitar', 'hard-rock', 'hardcore', 'hardstyle', 'heavy-metal',
       'hip-hop', 'honky-tonk', 'house', 'idm', 'indie-pop', 'indie',
       'industrial', 'j-dance', 'j-idol', 'j-pop', 'j-rock', 'jazz',
       'k-pop', 'malay', 'mandopop', 'metal', 'metalcore',
       'minimal-techno', 'mpb', 'new-age', 'opera', 'pagode', 'pop-film',
       'pop', 'power-pop', 'progressive-house', 'psych-rock', 'punk-rock',
       'punk', 'r-n-b', 'reggae', 'reggaeton', 'rock-n-roll', 'rock',
       'rockabilly', 'salsa', 'samba', 'sert

In [51]:
consolidated_spotify['track_genre'].nunique()

90

In [52]:
print(f'We now have 90 genres and {consolidated_spotify.shape[0]} songs in our dataset.')

We now have 90 genres and 62962 songs in our dataset.


After more research and observation, the following genres will also be removed: 
- malay
- groove
- guitar 
- pop film 

Malay is traditional malaysian music. This genre is similar to our langauge genres above where we saw a possible mix of traditional music with other genres already in the dataset. We can safely remove this one. 

Groove, guitar, and pop-film are all too general for our purposes. 

Let's remove these final genres. 

In [53]:
# creating list of column names
remaining_genre_cols = ['malay', 'groove', 'guitar', 'pop-film']

In [54]:
remaining_genres = consolidated_spotify[consolidated_spotify['track_genre'].isin(remaining_genre_cols)]

In [55]:
remaining_indexes = list(remaining_genres.index)

In [56]:
# Sanity check 
remaining_indexes

[43004,
 43008,
 43029,
 43040,
 43062,
 43064,
 43065,
 43067,
 43071,
 43087,
 43100,
 43107,
 43113,
 43114,
 43116,
 43119,
 43120,
 43121,
 43123,
 43124,
 43126,
 43130,
 43132,
 43134,
 43137,
 43138,
 43139,
 43140,
 43141,
 43143,
 43144,
 43156,
 43157,
 43160,
 43165,
 43166,
 43167,
 43168,
 43169,
 43170,
 43171,
 43172,
 43173,
 43174,
 43175,
 43176,
 43177,
 43180,
 43181,
 43182,
 43184,
 43185,
 43186,
 43187,
 43188,
 43189,
 43190,
 43191,
 43192,
 43193,
 43194,
 43195,
 43197,
 43198,
 43199,
 43204,
 43205,
 43206,
 43210,
 43211,
 43212,
 43213,
 43214,
 43217,
 43218,
 43219,
 43220,
 43223,
 43225,
 43226,
 43227,
 43228,
 43229,
 43230,
 43231,
 43232,
 43233,
 43234,
 43235,
 43236,
 43237,
 43238,
 43239,
 43240,
 43241,
 43243,
 43244,
 43245,
 43246,
 43247,
 43248,
 43249,
 43261,
 43265,
 43266,
 43268,
 43270,
 43271,
 43272,
 43274,
 43276,
 43277,
 43278,
 43279,
 43281,
 43282,
 43283,
 43284,
 43285,
 43286,
 43287,
 43289,
 43290,
 43291,
 43292,


In [57]:
consolidated_spotify.drop(remaining_indexes, inplace=True)

In [58]:
# Sanity check - our remaining genres
consolidated_spotify['track_genre'].unique()

array(['acoustic', 'afrobeat', 'alt-rock', 'alternative', 'ambient',
       'anime', 'black-metal', 'bluegrass', 'blues', 'breakbeat',
       'cantopop', 'chicago-house', 'classical', 'club', 'country',
       'dance', 'dancehall', 'death-metal', 'deep-house',
       'detroit-techno', 'disco', 'drum-and-bass', 'dub', 'dubstep',
       'edm', 'electro', 'electronic', 'emo', 'folk', 'forro', 'funk',
       'garage', 'gospel', 'goth', 'grindcore', 'grunge', 'hard-rock',
       'hardcore', 'hardstyle', 'heavy-metal', 'hip-hop', 'honky-tonk',
       'house', 'idm', 'indie-pop', 'indie', 'industrial', 'j-dance',
       'j-idol', 'j-pop', 'j-rock', 'jazz', 'k-pop', 'mandopop', 'metal',
       'metalcore', 'minimal-techno', 'mpb', 'new-age', 'opera', 'pagode',
       'pop', 'power-pop', 'progressive-house', 'psych-rock', 'punk-rock',
       'punk', 'r-n-b', 'reggae', 'reggaeton', 'rock-n-roll', 'rock',
       'rockabilly', 'salsa', 'samba', 'sertanejo', 'show-tunes',
       'singer-songwriter'

In [59]:
# Number of remaining genres
consolidated_spotify['track_genre'].nunique()

86

In [60]:
consolidated_spotify.shape

(59721, 20)

In [106]:
consolidated_spotify.to_csv('~/Desktop/CapstoneProject/data/consolidated_spotify_updated.csv', )

### NEXT STEPS

By removing all of these genres, we have lost a considerable amount of data, but there are ways to compensate for the smaller dataset. We also have the option of adding songs the same genres from our other dataset. The ambiguousness of these genres would have only hindered model performance. 

We still have a lot of genres that overlap with one another. For example, we can reduce everything down to the larger genres parent genres within the dataset - i.e. rock, pop, electronic, country, etc. Before we do any further consolidation of genres, let's run a few baseline models to see where we stand. 

***
**TABLE OF CONTENTS**

[Introduction](#Genre-Consolidation)

[Sleep](#Sleep)

[Study](#Study)

[Ambient](#Ambient)

[Sad & Happy](#Sad-&-Happy)

[Languages](#Languages)

[Next Steps](#NEXT-STEPS)

***

In [106]:
consolidated_spotify.to_csv('~/Desktop/CapstoneProject/data/consolidated_spotify_updated.csv', )